File: instrumentation/scala-fork-join-2.8/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/scalaexecutors/ForkJoinPoolBridge.java
/*
 * Copyright The OpenTelemetry Authors
 * SPDX-License-Identifier: Apache-2.0
 */

package io.opentelemetry.javaagent.instrumentation.scalaexecutors;

import java.util.Collection;
import java.util.List;
import java.util.concurrent.Callable;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Future;
import java.util.concurrent.TimeUnit;
import org.jetbrains.annotations.NotNull;
import scala.concurrent.forkjoin.ForkJoinPool;
import scala.concurrent.forkjoin.ForkJoinTask;

public class ForkJoinPoolBridge implements ExecutorService {

  private final ForkJoinPool delegate;

  public ForkJoinPoolBridge(ForkJoinPool delegate) {
    this.delegate = delegate;
  }

  @Override
  public void shutdown() {
    delegate.shutdown();
  }

  @NotNull
  @Override
  public List<Runnable> shutdownNow() {
    return delegate.shutdownNow();
  }

  @Override
  public boolean isShutdown() {
    return delegate.isShutdown();
  }

  @Override
  public boolean isTerminated() {
    return delegate.isTerminated();
  }

  @Override
  public boolean awaitTermination(long timeout, @NotNull TimeUnit unit)
      throws InterruptedException {
    return delegate.awaitTermination(timeout, unit);
  }

  @NotNull
  @Override
  public <T> Future<T> submit(@NotNull Callable<T> task) {
    return delegate.submit(task);
  }

  @NotNull
  @Override
  public <T> Future<T> submit(@NotNull Runnable task, T result) {
    return delegate.submit(task, result);
  }

  @NotNull
  @Override
  public Future<?> submit(@NotNull Runnable task) {
    return delegate.submit(task);
  }

  @NotNull
  @Override
  public <T> List<Future<T>> invokeAll(@NotNull Collection<? extends Callable<T>> tasks) {
    return delegate.invokeAll(tasks);
  }

  @NotNull
  @Override
  public <T> List<Future<T>> invokeAll(
      @NotNull Collection<? extends Callable<T>> tasks, long timeout, @NotNull TimeUnit unit) {
    // Scala 2.8's ForkJoinPool doesn't have corresponding method
    return delegate.invokeAll(tasks);
  }

  @NotNull
  @Override
  public <T> T invokeAny(@NotNull Collection<? extends Callable<T>> tasks)
      throws InterruptedException, ExecutionException {
    // Scala 2.8's ForkJoinPool doesn't have corresponding method
    return delegate.submit(tasks.iterator().next()).get();
  }

  @Override
  public <T> T invokeAny(
      @NotNull Collection<? extends Callable<T>> tasks, long timeout, @NotNull TimeUnit unit)
      throws InterruptedException, ExecutionException {
    // Scala 2.8's ForkJoinPool doesn't have corresponding method
    return delegate.submit(tasks.iterator().next()).get();
  }

  public <T> T invoke(ForkJoinTask<T> task) {
    return delegate.invoke(task);
  }

  @Override
  public void execute(@NotNull Runnable command) {
    delegate.execute(command);
  }

  public <T> void execute(ForkJoinTask<T> task) {
    delegate.execute(task);
  }
}

File: instrumentation/scala-fork-join-2.8/javaagent/src/test/scala/io/opentelemetry/javaagent/instrumentation/scalaexecutors/ForkJoinPoolBridge.scala
/*
 * Copyright The OpenTelemetry Authors
 * SPDX-License-Identifier: Apache-2.0
 */

package io.opentelemetry.javaagent.instrumentation.scalaexecutors

import java.util.concurrent._
import scala.concurrent.forkjoin.{ForkJoinPool, ForkJoinTask}

class ForkJoinPoolBridge(delegate: ForkJoinPool) extends ExecutorService {

  override def shutdown(): Unit = delegate.shutdown()

  override def shutdownNow(): java.util.List[Runnable] = delegate.shutdownNow()

  override def isShutdown: Boolean = delegate.isShutdown

  override def isTerminated: Boolean = delegate.isTerminated

  override def awaitTermination(timeout: Long, unit: TimeUnit): Boolean =
    delegate.awaitTermination(timeout, unit)

  override def submit[T](task: Callable[T]): Future[T] = delegate.submit(task)

  override def submit[T](task: Runnable, result: T): Future[T] =
    delegate.submit(task, result)

  override def submit(task: Runnable): Future[_] = delegate.submit(task)

  override def invokeAll[T](
      tasks: java.util.Collection[_ <: Callable[T]]
  ): java.util.List[Future[T]] =
    delegate.invokeAll(tasks)

  override def invokeAll[T](
      tasks: java.util.Collection[_ <: Callable[T]],
      timeout: Long,
      unit: TimeUnit
  ): java.util.List[Future[T]] =
    delegate.invokeAll(tasks)

  override def invokeAny[T](tasks: java.util.Collection[_ <: Callable[T]]): T =
    delegate.submit(tasks.iterator().next()).get()

  override def invokeAny[T](
      tasks: java.util.Collection[_ <: Callable[T]],
      timeout: Long,
      unit: TimeUnit
  ): T =
    delegate.submit(tasks.iterator().next()).get()

  def invoke[T](task: ForkJoinTask[T]): T = delegate.invoke(task)

  override def execute(command: Runnable): Unit = delegate.execute(command)

  def execute[T](task: ForkJoinTask[T]): Unit = delegate.execute(task)
}
